contrib/mesos/pkg/minion/task.go

```diff
@@ -0,0 +1,519 @@
+	t := makeTask(name, bin, args, env, cl)
+	go t.run(notStartedTask)
+	return t
+}
+
+// run executes the state machine responsible for starting, monitoring, and possibly restarting
+// a system process for the task. The initialState func is the entry point of the state machine.
+// Upon returning the done, completedCh, and errCh chans are all closed.
+func (t *task) run(initialState taskStateFn) {
+	t.startProcess = initialState
+
+	defer close(t.done)
+	defer close(t.completedCh)
+	defer close(t.errCh)
+
+	state := initialState
+	for state != nil {
+		next := state(t)
+		state = next
+	}
+}
+
+func (t *task) tryError(err error) {
+	select {
+	case <-t.shouldQuit:
+		// best effort
+		select {
+		case t.errCh <- err:
+		default:
+		}
+	case t.errCh <- err:
+	}
+}
+
+func (t *task) tryComplete(tc *taskCompletion) {
+	select {
+	case <-t.shouldQuit:
+		// best effort
+		select {
+		case t.completedCh <- tc:
+		default:
+		}
+	case t.completedCh <- tc:
+	}
+}
+
+type taskStateFn func(*task) taskStateFn
+
+func taskShouldRestart(t *task) taskStateFn {
+	select {
+	case <-t.shouldQuit:
+	default:
+		// sleep for a bit; then return t.startProcess
+		tm := time.NewTimer(t.restartDelay)
+		defer tm.Stop()
+		select {
+		case <-tm.C:
+			select {
+			case <-t.shouldQuit:
+			default:
+				if t.finished != nil {
+					if t.finished(true) {
+						return t.startProcess
+					}
+					// finish call decided not to respawn
+					return nil
+				}
+				return t.startProcess
+			}
+		case <-t.shouldQuit:
+		}
+	}
+	if t.finished != nil {
+		t.finished(false)
+	}
+	return nil
+}
+
+func (t *task) initLogging(r io.Reader) {
+	writer := t.createLogger()
+	go func() {
+		defer writer.Close()
+		_, err := io.Copy(writer, r)
+		if err != nil && err != io.EOF {
+			// using tryError is racy because the state machine closes errCh and
+			// so we don't want to attempt to write to a closed/closing chan. so
+			// log this for now.
+			log.Errorf("logger for task %q crashed: %v", t.bin, err)
+		}
+	}()
+}
+
+// notStartedTask spawns the given task and transitions to a startedTask state
+func notStartedTask(t *task) taskStateFn {
+	log.Infof("starting task process %q with args '%+v'", t.bin, t.args)
+
+	// create command
+	cmd := exec.Command(t.bin, t.args...)
+	if _, err := cmd.StdoutPipe(); err != nil {
+		// fatal error => terminate minion
+		t.tryError(fmt.Errorf("error getting stdout of %v: %v", t.name, err))
+		return taskShouldRestart
+	}
+	stderrLogs, err := cmd.StderrPipe()
+	if err != nil {
+		// fatal error => terminate minion
+		t.tryError(fmt.Errorf("error getting stderr of %v: %v", t.name, err))
+		return taskShouldRestart
+	}
+
+	t.initLogging(stderrLogs)
+	if len(t.env) > 0 {
+		cmd.Env = t.env
+	}
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid: true,
+	}
+
+	// last min check for shouldQuit here
+	select {
+	case <-t.shouldQuit:
+		return taskShouldRestart
```

Comment:
tried to centralize termination decisions in taskShouldRestart, instead of littering them all over the code